### PR TITLE
Drop tracking: handle invalid assignments better

### PR DESCRIPTION
**Pull request by eholk**
Previously this test case was crashing with an index out of bounds error deep in the call to `needs_drop`. We avoid this by detecting clearly invalid assignees in the `mutate` callback and ignoring these.

### CODE REVIEW

**Comment by compiler-errors** [comment]
@bors r-

wait one nit

File: compiler/rustc_typeck/src/check/generator_interior/drop_ranges/record_consumed_borrow.rs [hunk 1]
```diff
@@ -180,6 +180,15 @@ impl<'tcx> expr_use_visitor::Delegate<'tcx> for ExprUseDelegate<'tcx> {
         diag_expr_id: HirId,
     ) {
         debug!("mutate {assignee_place:?}; diag_expr_id={diag_expr_id:?}");
+
+        if assignee_place.place.base == PlaceBase::Rvalue
+            && assignee_place.place.projections.len() == 0
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/check/generator_interior/drop_ranges/record_consumed_borrow.rs:185]: nit: `is_empty` here maybe?

**Comment by compiler-errors** [comment]
@bors delegate+

sorry for the r+/r- whiplash, fix the nit or don't -- up to you :)

File: compiler/rustc_typeck/src/check/generator_interior/drop_ranges/record_consumed_borrow.rs (hunk 1 above)
Comment (eholk) [reply compiler/rustc_typeck/src/check/generator_interior/drop_ranges/record_consumed_borrow.rs]: Thanks, I like that way better too.

**Comment by eholk** [comment]
Thanks for the quick review!

@bors r=compiler-errors

**Pull request merged**